File: Trees/701. Insert into a Binary Search Tree/Solution.py
from collections import deque
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        # N : no of node in the Tree
        # Time Complexity: O(N)
        # Space Complexity: O(N) -> Stack Space
        # if not root:
        #     return TreeNode(val)

        # def dfs(root):
        #     if not root:
        #         return
            
        #     if root.val < val and not root.right:
        #         root.right = TreeNode(val)
        #     elif root.val > val and not root.left:
        #         root.left = TreeNode(val)
        #     elif root.val < val:
        #         dfs(root.right)
        #     elif root.val > val:
        #         dfs(root.left)
            
        # dfs(root)
        # return root


        # N : no of node in the Tree
        # Time Complexity: O(N)
        # Space Complexity: O(N)
        if not root:
            return TreeNode(val)

        stack = deque([root])
        while stack:
            node = stack.pop()
            if node.val < val and not node.right:
                node.right = TreeNode(val)
                return root
            elif node.val > val and not node.left:
                node.left = TreeNode(val)
                return root
            elif node.val < val:
                stack.append(node.right)
            elif node.val > val:
                stack.append(node.left)
        return root

File: Trees/701. Insert into a Binary Search Tree/Solution.java
import java.util.*;

//Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution {
    private void dfs(TreeNode root, int val) {
        if (root == null) return;

        if (root.val < val && root.right == null) root.right = new TreeNode(val);
        else if (root.val > val && root.left == null) root.left = new TreeNode(val);
        else if (root.val < val) dfs(root.right, val);
        else if (root.val > val) dfs(root.left, val);
        return; 
    }
    public TreeNode insertIntoBST(TreeNode root, int val) {
        // N : no of node in the Tree
        // Time Complexity: O(N)
        // Space Complexity: O(N) -> Stack Space
        // if (root == null) return new TreeNode(val);
        // dfs(root, val);
        // return root;


        // N : no of node in the Tree
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        if (root == null) return new TreeNode(val);
        
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            TreeNode node = stack.pop();
            if (node.val < val && node.right == null) {
                node.right = new TreeNode(val);
                return root;
            } else if (node.val > val && node.left == null) {
                node.left = new TreeNode(val);
                return root;
            } else if (node.val < val) stack.push(node.right);
            else if (node.val > val) stack.push(node.left);
        }
        return root;
    }
}